### PR TITLE
WTrackProperty: show simple editor on delayed double-click

### LINKED ISSUE #9981: enable editable track metadata in decks with 'selected click' lineedits
Reported by: **[ronso0](https://github.com/ronso0)**
Date: 2020-05-19T19:16:44Z
Status: Confirmed
Importance: Wishlist
Launchpad Issue: [lp1879553](https://bugs.launchpad.net/bugs/1879553)
Tags: library, skin, usability

---

Been dreaming of this ever sine I added the Comment row to decks in Tango skin:
* click track property to 'select' it (highlight background or frame)
* click again within N seconds to edit
* Enter to confirm, Esc to abort

related (fixed): Open track properties from decks lp:1166278

(tagged 'library' as it closely related like the track menu earlier)

### PR DESCRIPTION
**Pull request by ronso0**
show a simple QLineEdit after selected-click (delayed double click) onto track property labels

a coffee break project to help with my tag editing workflow (quick in-deck tag editing), i.e. create a fast alternative to double-click -> Track Properties -> click desired property line edit -> edit  -> hit Apply:

Gonna share this early, but Draft until it's polished.
track-label-edit-2

- [x] styles for line edit
- [x] ~~get real double-click interval, use x3 for selected click~~ 2s timeout are fine
- [x] use timer, reset 'selected' state automatically
- [x] visual feedback for 'selected' state (while widget expects second click):
new `selected` widget property, `WTrackProperty[selected="true"] { background: red; };`
- [x] add selected-click info to tooltips

May reuse #11752 for comments, but QLineEdit is fine for now.

### CODE REVIEW

**Comment by ronso0** [comment]
I'll probably keep polishing this, but it has already proven useful & usable in real life.

**Comment by ronso0** [comment]
This is very inviting, I edit and tag my tracks much more often now, and more spontaneously than before, when editing was rather clunky with Track Info or with the library.
(#11764 definitely helps, but this is way easier. )

Need to fix:
It's confusing that everytime an editor is closed the searchbar is focused. It should be the (library) widget that had focus before. Sending focus to tracks table is likely desired in the majority of cases, or a true re-focus.

**Comment by ronso0** [comment]
Ready to roll!

There's a new tr string for the respective tooltip.

**Comment by ronso0** [comment]
I added a workaround a Qt bug that became obvious with the new `hover` styles:
`hover` gets stuck when the track menu is opened which is kind of irritating. The hover state can only be reset by hovering the widget again after closing the menu.

Workaround is to post a QLeaveEvent when the menu is shown.

**Comment by ronso0** [comment]
rebased onto main

**Comment by ronso0** [comment]
> I added a workaround a Qt bug that became obvious with the new `hover` styles: `hover` gets stuck when the track menu is opened which is kind of irritating. The hover state can only be reset by hovering the widget again after closing the menu.
> 
> Workaround is to post a QLeaveEvent when the menu is shown.
> 
> Need to test whether this is still required with Qt6.

Yes, it is.

I also improved some comments. Ready for review.

**Comment by ronso0** [comment]
this is now based on #12004

**Comment by ronso0** [comment]
This works great, except: clicking outside the editor ('blank' area) doesn't close it. 
Related: https://github.com/mixxxdj/mixxx/issues/12966#issuecomment-1997106667)

**Comment by ronso0** [comment]
Very strange: after rebasing (main with new drag'n'drop threshold) hovering a WTrackProperty initiates a drag. THOUGH this is not caused by c++ changes but by qss selector `WTrackProperty:hover`.
🤷

File: src/widget/wtrackproperty.cpp [hunk 1]
```diff
@@ -2,15 +2,22 @@
...
 #include "moc_wtrackproperty.cpp"
 #include "skin/legacy/skincontext.h"
 #include "track/track.h"
 #include "util/dnd.h"
 #include "widget/wtrackmenu.h"
 
+namespace {
+// Duration (ms) the widget is 'selected' after left click, i.e. the duration
+// a second click would open the value editor
+constexpr int selectedClickTimeout = 2000;
```
Comment (daschuer) [inline src/widget/wtrackproperty.cpp:18]:
```suggestion
constexpr int kSelectedClickTimeoutMs = 2000;
```

File: src/widget/wtrackproperty.cpp [hunk 2]
```diff
@@ -74,21 +101,81 @@ void WTrackProperty::slotTrackChanged(TrackId trackId) {
...
+
+    // Don't create the editor or toggle the 'selected' state for protected
+    // properties like duration.
+    if (!m_propertyIsWritable) {
+        return;
+    }
+
+    if (!m_pSelectedClickTimer) {
+        // create & start the timer
+        m_pSelectedClickTimer = new QTimer(this);
```
Comment (daschuer) [inline src/widget/wtrackproperty.cpp:145]: This can also become a parented pointer.

**Review by daschuer — CHANGES_REQUESTED** [review]
Works good. I have only two minor comments

**Review by daschuer — APPROVED** [review]
LGTM, Thank you.

**Comment by ronso0** [comment]
I squashed the fixup.

**Comment by daschuer** [comment]
Thank you.

**Comment by ronso0** [comment]
Wooohoo :tada: 
Thanks for reviewing @daschuer

**Comment by ronso0** [comment]
Review after having used this for a few months:
quite often I expected the editor to open after the first click (select), probably due to the hover effect.
I'll check if I can make it act on first click and while leaving drag'n'drop as is.

**Pull request merged**